### PR TITLE
Support load aware scheduling

### LINKED ISSUE #95: [proposal] scheduler support node load balancing
**What would you like to be added**:
Support scheduling based on node load. Nodes with low load are allocated first, and when the load reaches a safe threshold, consideration should be given not to continue allocation. 

We can also consider supporting the ability of portraits to describe the resource usage of different workloads, and to guide the scheduler to achieve finer load balancing through portrait data.

By the way, in the actual production environment, the load of the node will not increase quickly and can be quickly perceived by the scheduler. In the case of no portrait, we should consider providing a mechanism to deal with this scenario.

**Why is this needed**:

Koordinator allowing for resource overcommitments to achieve high resource utilizations. We must reduce the impact of co-location on latency-sensitive applications and improve runtime reliability.

### PR DESCRIPTION
**Pull request by eahydra**
### Ⅰ. Describe what this PR does

The scheduling plugin filters abnormal nodes and scores them according
to resource usage. The plugin extends the Filter/Score/Reserve/Unreserve
extension points defined in the Kubernetes scheduling framework.

FYI: [docs/proposals/scheduling/20220510-load-aware-scheduling.md](https://github.com/koordinator-sh/koordinator/blob/main/docs/proposals/scheduling/20220510-load-aware-scheduling.md)

### Ⅱ. Does this pull request fix one issue?

Fix #95 

### Ⅲ. Describe how to verify it

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta2
kind: KubeSchedulerConfiguration
profiles:
- pluginConfig:
  - args:
      apiVersion: kubescheduler.config.k8s.io/v1beta2
      kind: LoadAwareSchedulingArgs
      filterExpiredNodeMetrics: true
      nodeMetricExpirationSeconds: 300
      resourceWeights:
        cpu: 2
        memory: 1
      usageThresholds:
        cpu: 75
        memory: 85
      estimatedScalingFactors:
        cpu: 80
        memory: 70
    name: LoadAwareScheduling
  plugins:
    filter:
      enabled:
        - name: LoadAwareScheduling
          weight: 0
    reserve:
      enabled:
        - name: LoadAwareScheduling
          weight: 0
    score:
      enabled:
        - name: LoadAwareScheduling
          weight: 1000
  schedulerName: koord-scheduler
```

### CODE REVIEW

File: hack/update-scheduler-codegen.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+SCRIPT_ROOT=$(dirname "${BASH_SOURCE[@]}")/..
+
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
+
+bash "${CODEGEN_PKG}"/generate-internal-groups.sh \
+  "deepcopy,conversion,defaulter" \
+  github.com/koordinator-sh/koordinator/pkg/scheduler/apis/generated \
+  github.com/koordinator-sh/koordinator/pkg/scheduler/apis \
+  github.com/koordinator-sh/koordinator/pkg/scheduler/apis \
+  "config:v1beta2" \
```
[the remark this answers is not in the document]
Comment (eahydra) [reply hack/update-scheduler-codegen.sh:19]: k8s scheduling framework constraints the api version... If the framework configuration use v1beta2, the configuration version of all plugins must also be v1beta2...it's so amazing

File: pkg/scheduler/apis/config/types_pluginargs.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// LoadAwareSchedulingArgs holds arguments used to configure the LoadAwareScheduling plugin.
+type LoadAwareSchedulingArgs struct {
+	metav1.TypeMeta
+
+	FilterExpiredNodeMetrics    *bool                         `json:"filterExpiredNodeMetrics,omitempty"`
```
Comment (jasonliu747) [inline pkg/scheduler/apis/config/types_pluginargs.go:30]: Add a comment to each field. Just copy and paste from your proposal 😏

File: pkg/scheduler/apis/config/v1beta2/defaults.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+package v1beta2
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/pointer"
+)
+
+var (
+	defaultNodeMetricExpirationSeconds int64 = 180
```
Comment (jasonliu747) [inline pkg/scheduler/apis/config/v1beta2/defaults.go:25]: can you add some comments to briefly tell us abouth the stroy behind these default values?

File: pkg/util/pod.go [hunk 4]
```diff
@@ -243,3 +243,7 @@ func GetPodRequest(pod *corev1.Pod, resourceNames ...corev1.ResourceName) corev1
 	}
 	return result
 }
+
+func TerminatedPod(pod *corev1.Pod) bool {
```
Comment (jasonliu747) [inline pkg/util/pod.go:247]: this function name is a little bit confusing, I would assume it terminates a pod.

**Review by jasonliu747 — COMMENTED** [review]
haven't gone through the core logic of load aware plugin implementation yet. here is what I got for you so far ;)

File: pkg/scheduler/plugins/loadaware/load_aware.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+			return DefaultMilliCPURequest
+		case corev1.ResourceMemory, extension.BatchMemory:
+			return DefaultMemoryRequest
+		}
+		return 0
+	}
+
+	var estimatedUsed int64
+	switch resourceName {
+	case corev1.ResourceCPU:
```
Comment (hormes) [inline pkg/scheduler/plugins/loadaware/load_aware.go:260]: If it is a besteffort Pod？

File: cmd/koord-scheduler/app/options/options.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package options
+
+import (
+	scheduledoptions "k8s.io/kubernetes/cmd/kube-scheduler/app/options"
```
[the remark this answers is not in the document]
Comment (eahydra) [reply cmd/koord-scheduler/app/options/options.go:20]: done

File: pkg/scheduler/apis/config/doc.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+// +k8s:deepcopy-gen=package
+// +groupName=scheduler.config.koordinator.sh
+
+// Package config of the API.
+package config // import "github.com/koordinator-sh/koordinator/pkg/scheduler/apis/config"
```
[the remark this answers is not in the document]
Comment (eahydra) [reply pkg/scheduler/apis/config/doc.go:21]: done

File: pkg/scheduler/apis/config/types_pluginargs.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
+/*
```
[the remark this answers is not in the document]
Comment (eahydra) [reply pkg/scheduler/apis/config/types_pluginargs.go:1]: done

File: pkg/scheduler/apis/config/types_pluginargs.go (hunk 2 above)
Comment (eahydra) [reply pkg/scheduler/apis/config/types_pluginargs.go:30]: done

File: pkg/util/pod.go (hunk 4 above)
Comment (eahydra) [reply pkg/util/pod.go:247]: use `IsPodTerminated`.

File: pkg/scheduler/plugins/loadaware/pod_assign_cache.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	lock         sync.RWMutex
+	podInfoItems map[string]map[types.UID]*podAssignInfo
+}
+
+type podAssignInfo struct {
+	timestamp time.Time
+	pod       *corev1.Pod
+}
+
+func newPodAssignCache() *podAssignCache {
```
[the remark this answers is not in the document]
Comment (eahydra) [reply pkg/scheduler/plugins/loadaware/pod_assign_cache.go:44]: From the past scheduling function development experience, most of the data is only used by a certain plugin. The overall resource cost is acceptable

File: pkg/scheduler/plugins/loadaware/pod_assign_cache.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/koordinator-sh/koordinator/pkg/util"
+)
+
+var (
+	timeNowFn = time.Now
+)
+
+type podAssignCache struct {
+	lock         sync.RWMutex
+	podInfoItems map[string]map[types.UID]*podAssignInfo
```
[the remark this answers is not in the document]
Comment (eahydra) [reply pkg/scheduler/plugins/loadaware/pod_assign_cache.go:36]: done

File: pkg/scheduler/plugins/loadaware/load_aware.go (hunk 5 above)
Comment (eahydra) [reply pkg/scheduler/plugins/loadaware/load_aware.go:260]: The best-effort Pod uses extended resources and can be calculated directly using Value().

**Comment by eahydra** [comment]
support user-defined node resource utilization thresholds.

File: pkg/scheduler/plugins/loadaware/load_aware.go [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+	return nodeMetric == nil ||
+		nodeMetric.Status.UpdateTime == nil ||
+		nodeMetricExpirationSeconds > 0 &&
+			time.Since(nodeMetric.Status.UpdateTime.Time) >= time.Duration(nodeMetricExpirationSeconds)*time.Second
+}
+
+func (p *Plugin) estimatedAssignedPodUsage(nodeName string, nodeMetric *slov1alpha1.NodeMetric) map[corev1.ResourceName]int64 {
+	estimatedUsed := make(map[corev1.ResourceName]int64)
+	nodeMetricReportInterval := getNodeMetricReportInterval(nodeMetric)
+	p.podAssignCache.lock.RLock()
```
Comment (saintube) [inline pkg/scheduler/plugins/loadaware/load_aware.go:215]: Is it possible to reduce the lock granularity as there is no correlation among different nodes?

File: pkg/scheduler/plugins/loadaware/load_aware.go [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+
+	var estimatedUsed int64
+	switch resourceName {
+	case corev1.ResourceCPU:
+		estimatedUsed = int64(math.Round(float64(quantity.MilliValue()) * float64(scalingFactor) / 100))
+		if estimatedUsed > limitQuantity.MilliValue() {
+			estimatedUsed = limitQuantity.MilliValue()
+		}
+	default:
+		estimatedUsed = int64(math.Round(float64(quantity.Value()) * float64(scalingFactor) / 100))
```
Comment (saintube) [inline pkg/scheduler/plugins/loadaware/load_aware.go:277]: How about keeping the type `resource.Quantity` until the final scoring to improve precisions?

File: pkg/scheduler/plugins/loadaware/load_aware.go (hunk 11 above)
Comment (saintube) [reply pkg/scheduler/plugins/loadaware/load_aware.go:215]: ok

**Comment by saintube** [comment]
/lgtm  
others PTAL

**Pull request merged**